### PR TITLE
Add Bezier curve primitive

### PR DESCRIPTION
**Pull request by serenity4**
Adds a new `BezierCurve` type, which can be evaluated by calling on its parametric variable. I implemented two methods, namely Horner's method and De Casteljau's algorithm, that can be used to evaluate it. The first is a fast approximation, useful with many control points, while the second is very accurate but also more expensive. I made the second one be the default, because with few control points (I tested with 3) the performance is very similar.

### CODE REVIEW

File: src/primitives/bezier.jl
Comment (serenity4) [inline src/primitives/bezier.jl]: I wasn't sure whether I should define a function for this or not. The idea was not to copy paste this for the two algorithms. The first argument type is just used for dispatch. If that makes sense, I can extend this function with other methods for other types which have a restricted domain (all callable parametric types for now), in a separate PR. To have atomic changes I could just remove this here, copy-paste the error handling code, and recreate this function in a separate PR, if you deem it preferable.

File: src/primitives/bezier.jl
Comment (serenity4) [inline src/primitives/bezier.jl:38]: This abstract type is not concretely used, but I found it was clearer to appropriately subtype both method types.

**Comment by serenity4** [comment]
The code for Horner's method was adapted from [Bezier.jl](https://github.com/arlk/Bezier.jl). Since this is essentially a copy-paste with minor modifications, should I specify it somewhere? Maybe in the dosctring? The package is MIT-licensed.

File: src/primitives/bezier.jl
Comment (serenity4) [inline src/primitives/bezier.jl]: I took the same notation as the author's, but I don't know where it comes from. Should I change it? This one has the advantage of being concise, but without context I admit it is difficult to decipher it.

File: src/primitives/bezier.jl
Comment (juliohm) [reply src/primitives/bezier.jl]: Yes, maybe we should introduce this helper function in a separate PR and then add methods to it for the different geometries like Segment, Ray and now Bezier :)

File: src/primitives/bezier.jl
Comment (juliohm) [inline src/primitives/bezier.jl]: Why can't we use `BezierCurve{Dim,T}` here?

File: src/primitives/bezier.jl
Comment (juliohm) [inline src/primitives/bezier.jl]: Perhaps this `coordinates(last(cs))` could be moved up in the function body near the `coordinates(first(cs))` ? Perhaps these two could be computed together?

File: src/primitives/bezier.jl
Comment (juliohm) [reply src/primitives/bezier.jl]: It would be nice if we could find a good reference to cite here, and then follow the same notation there. It is difficult to know what the code is doing otherwise, but it is indeed very elegant.

File: src/primitives/bezier.jl
Comment (juliohm) [inline src/primitives/bezier.jl:9]: From the Wikipedia article, what is the type of Bezier curve you implemented? Is it of a fixed order (linear, quadratic, cubic) or a recursive definition? I need to learn about these curves more seriously at some point. They are really nice.

File: src/primitives/bezier.jl
Comment (serenity4) [reply src/primitives/bezier.jl]: We can, it's just that `Dim` won't be used, so there is no need to request the parameter.

File: src/primitives/bezier.jl
Comment (serenity4) [reply src/primitives/bezier.jl:9]: It's a recursive definition, for both evaluation algorithms. If you have a curve with 3 points it's quadratic, 4 it's cubic, and so on.

File: src/primitives/bezier.jl
Comment (juliohm) [reply src/primitives/bezier.jl]: Let's use Dim explicitly then, it makes the code more readable.

File: src/primitives/bezier.jl
Comment (juliohm) [reply src/primitives/bezier.jl:9]: I would state this explicitly in the doc string then: `A recursive Bezier curve with control points`.

File: src/primitives/bezier.jl
Comment (serenity4) [reply src/primitives/bezier.jl]: The Wikipedia page + the new docstring should be enough. Actually, it's not complicated at all when you introduce context.

**Comment by serenity4** [comment]
I ended up refactoring the evaluation with Horner's method, at such point that it does not look like the original version at all. It is more natural, follows Wikipedia's notations and is 3x faster as a side-effect.

File: src/primitives/bezier.jl
Comment (serenity4) [inline src/primitives/bezier.jl:101]: I would agree that this is useless, but it makes it clearer when comparing to the original algorithm.

File: src/primitives/bezier.jl
Comment (serenity4) [inline src/primitives/bezier.jl]: `bn` is not used at all, but it adds clarity when looking at the notations (cf docstring). And, thanks to dead code elimination, there is no performance loss at all when doing that (I checked).

**Comment by serenity4** [comment]
Because Horner's method is now faster, there is a 3x performance difference when evaluating a Bezier curve even for a quadratic one (3 points). So, either we provide accuracy with OK performance as a default, or a fast approximation. I still think we should prioritize accuracy, so that people don't encounter dark bugs when we can avoid it, and if they want speed then they are warned about stability issues and make a conscious choice.

File: src/primitives/bezier.jl
Comment (juliohm) [inline src/primitives/bezier.jl]:
```suggestion
BezierCurve(points::AbstractVector{<:Tuple}) = BezierCurve(Point.(points))
BezierCurve(points::Vararg) = BezierCurve(collect(points))
```

File: src/primitives/bezier.jl
Comment (juliohm) [inline src/primitives/bezier.jl]: I think we don't need to collect the segments here, right?

Comment (serenity4) [reply src/primitives/bezier.jl]: Ooh, right.

File: src/primitives/bezier.jl
Comment (serenity4) [inline src/primitives/bezier.jl]:
```suggestion
  ss = segments(Chain(curve.controls))
```

File: src/primitives/bezier.jl
Comment (juliohm) [inline src/primitives/bezier.jl:77]: Love these docstrings ❤️

File: src/primitives/bezier.jl
Comment (juliohm) [inline src/primitives/bezier.jl:32]: Just double checking, degree is the most accurate term right? It is not order or something else...

File: src/primitives/bezier.jl
Comment (juliohm) [inline src/primitives/bezier.jl]: Can you explain this `@views` along with the coordinates of a single point? Also, maybe we should use `last(cs)` to be consistent with other access patterns in the codebase?

File: src/primitives/bezier.jl
Comment (juliohm) [reply src/primitives/bezier.jl]: Is this code elimination guaranteed or something that the compiler may fail in a future release? I like code clarity but unused seem more annoying. Perhaps just a comment instead of introducing a variable name would be better. Can we elaborate on the comment instead?

File: src/primitives/bezier.jl
Comment (juliohm) [reply src/primitives/bezier.jl:101]: At least this one is being used later in the `Point` return. 👍🏽

File: test/runtests.jl
Comment (juliohm) [inline test/runtests.jl]: Should we add this constant definition inside the testset? That way we don't need to keep track of the main runtests.jl file when things change in the future.

**Review by juliohm — CHANGES_REQUESTED** [review]
This PR will be super useful for a tracing algorithm I am currently implementing ❤️ Let's just fix the minor details before we merge it.

**Comment by juliohm** [comment]
> Because Horner's method is now faster, there is a 3x performance difference when evaluating a Bezier curve even for a quadratic one (3 points). So, either we provide accuracy with OK performance as a default, or a fast approximation. I still think we should prioritize accuracy, so that people don't encounter dark bugs when we can avoid it, and if they want speed then they are warned about stability issues and make a conscious choice.

Yes, I think it is a reasonable choice. Make the accurate method the default and then use the approximation only as an option. 👍🏽

File: src/primitives/bezier.jl
Comment (serenity4) [reply src/primitives/bezier.jl:32]: Hmm I actually find both terms to be used. Although I quickly googled "order vs degree", and found that the order of a polynomial is the number of terms in a series expansion, and so it equals the degree + 1. So, here it is the degree, and `ncontrols` actually equals the order.

File: src/primitives/bezier.jl
Comment (serenity4) [reply src/primitives/bezier.jl]: I agree for `last(cs)`, will a view work with this though? And, I can replace `@views` with `@view`, yes.

File: src/primitives/bezier.jl
Comment (serenity4) [reply src/primitives/bezier.jl]: The compiler should never fail such a basic optimization. I mean, this would be a serious problem that I think the developers won't let happen.

File: src/primitives/bezier.jl
Comment (juliohm) [reply src/primitives/bezier.jl:32]: Yes, just wanted to double check. I think degree is perfectly fine 💯

File: src/primitives/bezier.jl
Comment (serenity4) [reply src/primitives/bezier.jl]: Well, I removed the view there, I don't think it actually makes a difference for just 1 point.

**Comment by serenity4** [comment]
I think it is ready to merge :100:

**Review by juliohm — APPROVED** [review]

**Comment by juliohm** [comment]
That is an amazing PR. Thanks! Let's just wait for the tests to pass before we merge.

**Pull request merged**
